File: src/lib/audio/effects/creative/ring-modulator.ts
// Ring Modulator Effect Processor
// Carrier frequency multiplication for metallic/robotic tones
// Zero additional latency - uses native oscillator multiplication

import { BaseEffect } from '../base-effect';
import type { RingModulatorSettings } from '@/types';

export class RingModulatorProcessor extends BaseEffect {
  readonly name = 'Ring Modulator';

  private settings: RingModulatorSettings;

  // Carrier oscillator
  private carrier: OscillatorNode;
  private carrierGain: GainNode;

  // Ring modulation is achieved by multiplying input with carrier
  // In Web Audio, we use a gain node with the carrier as gain modulator

  // Dry/wet mixing
  private dryGain: GainNode;
  private modGain: GainNode;

  // Envelope follower for FM (optional)
  private envelopeFilter: BiquadFilterNode;

  constructor(audioContext: AudioContext, settings?: Partial<RingModulatorSettings>) {
    super(audioContext);

    this.settings = {
      enabled: false,
      frequency: 440, // Hz (20-5000)
      waveform: 'sine', // sine, square, triangle, sawtooth
      mix: 50, // 0-100
      lfoRate: 0, // 0-10 Hz (0 = no LFO)
      lfoDepth: 0, // 0-100
      ...settings,
    };

    // Create carrier oscillator
    this.carrier = audioContext.createOscillator();
    this.carrier.type = this.settings.waveform;
    this.carrier.frequency.value = this.settings.frequency;

    // Carrier gain (the modulator)
    this.carrierGain = audioContext.createGain();
    this.carrierGain.gain.value = 0; // Will be modulated by input

    // Connect carrier to gain control
    this.carrier.connect(this.carrierGain.gain);

    // Dry/wet
    this.dryGain = audioContext.createGain();
    this.modGain = audioContext.createGain();

    // Envelope follower for carrier AM
    this.envelopeFilter = audioContext.createBiquadFilter();
    this.envelopeFilter.type = 'lowpass';
    this.envelopeFilter.frequency.value = 50;

    // Wire up signal chain
    // Dry path
    this.inputGain.connect(this.dryGain);
    this.dryGain.connect(this.wetGain);

    // Ring modulation path
    // Input goes through carrierGain which is modulated by carrier oscillator
    this.inputGain.connect(this.carrierGain);
    this.carrierGain.connect(this.modGain);
    this.modGain.connect(this.wetGain);

    this.wetGain.connect(this.outputGain);

    // Start carrier
    this.carrier.start();

    // Apply initial settings
    this.updateFrequency();
    this.updateMix();
  }

  private updateFrequency(): void {
    const now = this.audioContext.currentTime;
    this.carrier.frequency.setTargetAtTime(this.settings.frequency, now, 0.01);
  }

  private updateWaveform(): void {
    this.carrier.type = this.settings.waveform;
  }

  private updateMix(): void {
    const now = this.audioContext.currentTime;
    const mix = this.settings.mix / 100;

    // Equal power crossfade
    this.dryGain.gain.setTargetAtTime(Math.sqrt(1 - mix), now, 0.01);
    this.modGain.gain.setTargetAtTime(Math.sqrt(mix), now, 0.01);
  }

  updateSettings(settings: Partial<RingModulatorSettings>): void {
    this.settings = { ...this.settings, ...settings };

    if (settings.enabled !== undefined) {
      this.setEnabled(settings.enabled);
    }

    if (settings.frequency !== undefined) {
      this.updateFrequency();
    }

    if (settings.waveform !== undefined) {
      this.updateWaveform();
    }

    if (settings.mix !== undefined) {
      this.updateMix();
    }
  }

  getSettings(): RingModulatorSettings {
    return { ...this.settings };
  }

  dispose(): void {
    this.carrier.stop();
    this.carrier.disconnect();
    this.carrierGain.disconnect();
    this.dryGain.disconnect();
    this.modGain.disconnect();
    this.envelopeFilter.disconnect();
    super.dispose();
  }
}

File: native-bridge/src/effects/ring_modulator.rs
//! Ring Modulator - carrier frequency multiplication

use super::dsp::{Lfo, LfoWaveform};
use super::types::{RingModWaveform, RingModulatorSettings};
use super::AudioEffect;
use std::f32::consts::TAU;

pub struct RingModulator {
    settings: RingModulatorSettings,
    sample_rate: f32,
    carrier_phase: f32,
    lfo: Lfo,
}

impl RingModulator {
    pub fn new(sample_rate: u32) -> Self {
        Self {
            settings: RingModulatorSettings::default(),
            sample_rate: sample_rate as f32,
            carrier_phase: 0.0,
            lfo: Lfo::new(LfoWaveform::Sine),
        }
    }

    pub fn update_settings(&mut self, settings: RingModulatorSettings) {
        let waveform = match settings.waveform {
            RingModWaveform::Sine => LfoWaveform::Sine,
            RingModWaveform::Triangle => LfoWaveform::Triangle,
            RingModWaveform::Square => LfoWaveform::Square,
            RingModWaveform::Sawtooth => LfoWaveform::Sawtooth,
        };
        self.lfo.set_waveform(waveform);
        self.settings = settings;
    }

    pub fn set_sample_rate(&mut self, rate: u32) {
        self.sample_rate = rate as f32;
    }

    /// Generate carrier waveform sample
    #[inline]
    fn carrier(&self, phase: f32) -> f32 {
        match self.settings.waveform {
            RingModWaveform::Sine => (phase * TAU).sin(),
            RingModWaveform::Triangle => {
                let p = phase % 1.0;
                if p < 0.25 {
                    p * 4.0
                } else if p < 0.75 {
                    2.0 - p * 4.0
                } else {
                    p * 4.0 - 4.0
                }
            }
            RingModWaveform::Square => {
                if phase % 1.0 < 0.5 {
                    1.0
                } else {
                    -1.0
                }
            }
            RingModWaveform::Sawtooth => 2.0 * (phase % 1.0) - 1.0,
        }
    }
}

impl AudioEffect for RingModulator {
    fn process(&mut self, samples: &mut [f32], _sample_rate: u32) {
        if !self.settings.enabled {
            return;
        }

        let base_freq = self.settings.frequency;
        let lfo_rate = self.settings.lfo_rate;
        let lfo_depth = self.settings.lfo_depth / 100.0;
        let mix = self.settings.mix / 100.0;

        for frame in samples.chunks_mut(2) {
            // Get LFO modulation for carrier frequency
            let lfo_val = if lfo_rate > 0.0 {
                self.lfo.tick(lfo_rate, self.sample_rate)
            } else {
                0.0
            };

            // Modulated carrier frequency
            let freq = base_freq * (1.0 + lfo_val * lfo_depth);

            // Generate carrier
            let carrier = self.carrier(self.carrier_phase);

            // Advance carrier phase
            self.carrier_phase = (self.carrier_phase + freq / self.sample_rate) % 1.0;

            // Ring modulation: multiply signal by carrier
            let dry_left = frame[0];
            let mod_left = dry_left * carrier;

            frame[0] = dry_left * (1.0 - mix) + mod_left * mix;

            if frame.len() > 1 {
                let dry_right = frame[1];
                let mod_right = dry_right * carrier;
                frame[1] = dry_right * (1.0 - mix) + mod_right * mix;
            }
        }
    }

    fn reset(&mut self) {
        self.carrier_phase = 0.0;
        self.lfo.reset();
    }

    fn is_enabled(&self) -> bool {
        self.settings.enabled
    }

    fn set_enabled(&mut self, enabled: bool) {
        self.settings.enabled = enabled;
    }
}
